Add Ubuntu build automation via PNNL GitLab

jacksavage opened the pull request:
- adds `install_package_deps_lib.sh` to install packages to each container image based on the detected distribution.
- updates `install_gridpack.sh`.
- updates `install_from_source_deps.sh` (was `install_gridpack_deps.sh`) and breaks out pieces into `install_boost.sh`, `install_ga.sh`, and `install_petsc.sh`.
- adds definition for a gitlab build pipeline via `.gitlab-ci.yml`. this contains 1) jobs to test if the container image needs rebuilt, 2) a job to build a container image used as the environment for subsequent jobs, 3) a job that installs gridpack, and 4) a job that runs the tests for that gridpack installation.
- adds `dockerfile` which is the definition for the container image to be built. besides setting environment variables, this uses `install_boost.sh`, `install_ga.sh`, `install_petsc.sh`, and `install_gridpack_deps.sh`.
- split off of #173 to get the ubuntu build incorporated while troubleshooting continues on the rocky linux build

abhyshr committed:
Added script to build GridPACK in one-go, including installing all dependencies.

abhyshr committed:
Updated installation script tested on EIOC-Ubuntu8

abhyshr committed:
Minor modifications to install script

abhyshr committed:
Small modification to check python

abhyshr committed:
Updated gridpack install script to only install GridPACK and python wrapper.

abhyshr committed:
More work on installation script

- Split the script into two (one for installing dependencies and other for GridPACK)
- Ignore boost includes and boost libraries in GridPACK.cmake.in. Otherwise python
  installation fails.
- Update GridPACK installation script with addition CMake flags for Boost.

abhyshr committed:
Create external-dependencies directory, if not present.

wperkins committed:
Remove GA-based dense matrix; adjust matrix transpose

wperkins committed:
Remove all PETSC_VERSION usage.

wperkins committed:
Update PETSc settings for tlaloc

wperkins committed:
Add smoke test for 240-bus dynamic simulation

wperkins committed:
Add a CMake macro to do the LU solver check and use it

wperkins committed:
Add smoke test for 2-area dynamic simulation

wperkins committed:
Add smoke test for two area case w/ renewables

wperkins committed:
Remove math::transpose(Matrix<>, Matrix<>) because of PETSc transpose semantics

wperkins committed:
Add smoke tests for dsf2.x (same as for dsf.x).

wperkins committed:
Add smoke test for Wind DSA 9-bus case

wperkins committed:
Fix typos in Wind DSA TAMU 2000 case

wperkins committed:
Add a smoke test for Wind DSA TAMU 2000 case

jacksavage committed:
Apply shellcheck suggestions and run formatter

jacksavage committed:
Move blocks into functions and extract versions

- Remove directory changes and assume current directory
- Use local variables in functions
- Capture logs

jacksavage committed:
Correct how variables are checked

jacksavage committed:
Bump versions out of script with env vars

wperkins committed:
Remove ancient test from `applications/development/dynamic_simulation_reduced_y`

jacksavage committed:
Fix env var

jacksavage committed:
Set number of make threads dynamically

jacksavage committed:
Add ci script and container image definition

wperkins committed:
Fix incorrect Matrix type Jacobian supplied to DAESolver constructor

wperkins committed:
Update my build script

jacksavage committed:
Test pipeline trigger

jacksavage committed:
Build container, make dep from test optional

jacksavage committed:
Update dockerfile path

jacksavage committed:
Increase k8s memory limit

jacksavage committed:
Increase mem request/limit and job timeout

jacksavage committed:
Add cpu and increase timeout further

jacksavage committed:
Troubleshooting env

jacksavage committed:
Use correct predefined CI variable

jacksavage committed:
Add conditions for building container

And enable real test job

jacksavage committed:
Adjust k8s parameters and gitlab job timeout

jacksavage committed:
Update CPU limit

jacksavage committed:
Adjust CPU req/limits

jacksavage committed:
Adjust CPU limit

jacksavage committed:
Adjust cpu request

jacksavage committed:
Set GRIDPACK_DIR

jacksavage committed:
Uncomment gp_include and gp_libs

jacksavage committed:
Add Boost_NO_BOOST_CMAKE

jacksavage committed:
Set max number of make jobs

jacksavage committed:
Only append python path if exists

jacksavage committed:
Add comment

jacksavage committed:
Add test job

jacksavage committed:
Use default timeout for build

wperkins committed:
Remove 2000-bus Wind-DSA smoke tests - they take too long

wperkins committed:
Switch to a direct linear solver because this was diverging

jacksavage committed:
Whitespace change

jacksavage committed:
Merge remote-tracking branch 'origin/fix/testing' into gitlab-ci

jacksavage committed:
Troubleshooting

jacksavage committed:
Revert block comment

jacksavage committed:
Fix test command and save results

jacksavage committed:
Always save artifacts for the test job

jacksavage committed:
Remove version numbers from LD_LIBRARY_PATH

jacksavage committed:
Run tests as non-root user

jacksavage committed:
Fix rules for build-container job

jacksavage committed:
Don't override env vars

jacksavage committed:
Remove manual container build rule

jacksavage committed:
Add build status badge

jacksavage committed:
Only use two processors

jacksavage committed:
Show output for failed tests

jacksavage committed:
Test with multiple distributions

jacksavage committed:
Merge branch 'develop' into gitlab-ci

jacksavage committed:
Back out changes to src

jacksavage committed:
Remove attempted variable substitution

jacksavage committed:
Set conditions for when container is built

jacksavage committed:
Fix quoting

jacksavage committed:
Make script available to setup env

jacksavage committed:
Make scripts executable for docker image build

jacksavage committed:
Save logs from install_gridpack_deps

jacksavage committed:
Fix artifact path for logs

jacksavage committed:
Test - remove parallel config

jacksavage committed:
Link logs to project folder

- artifacts can only be saved from there per gitlab docs

jacksavage committed:
Send logs to back console 😫

jacksavage committed:
Redirect output to files

jacksavage committed:
Build PETSc with MUMPS instead of SuperLU_dist

jacksavage committed:
Redirect both streams from kaniko exe to file

Still not getting an artifact in gitlab... not sure if it's bc I'm overflowing the console

jacksavage committed:
Move the container registry check to its own job

with curl actually installed

jacksavage committed:
Use block scalar to avoid issues with special chars

jacksavage committed:
Always save untracked files

jacksavage committed:
Fix block scalar

Lines with indentation are not folded! https://yaml-multiline.info/

jacksavage committed:
Swap out mpms for superlu for now

PETSc config fails due to options and I need to figure out a way to get the more detailed log from outside the project folder (as a gitlab artifact).

jacksavage committed:
Pull back on untracked artifacts where it makes sense

jacksavage committed:
Revert "Swap out mpms for superlu for now"

This reverts commit 9e2637071f4b875136f309ea5c1eb1f40507c541.

jacksavage committed:
Add scalapack per error message

jacksavage committed:
Remove the artifact path spec for build container

For some reason this wasn't saved again...

jacksavage committed:
Pull container repo id into var

jacksavage committed:
Try to see log files I need when build fails

jacksavage committed:
Set the proxy var for kaniko

jacksavage committed:
Pass proxy to all jobs

jacksavage committed:
Try lowercase

jacksavage committed:
Mark as variables

jacksavage committed:
Remove petsc config option

jacksavage committed:
Remove unused log folder

jacksavage committed:
Try to parallelize the pipeline again

Do everything for ubuntu and rockylinux

jacksavage committed:
Revert troubleshooting command

jacksavage committed:
Separate scripts to improve caching

jacksavage committed:
Install mpi4py dependency

jacksavage committed:
Set proxy build args in container build env

jacksavage committed:
Get mpi4py from system repos instead of through pip

jacksavage committed:
Use common check for distro between scripts

jacksavage committed:
Manually define job parallelization and adjust whitespace

jacksavage committed:
Try adding docker config and cert to kaniko container

Also disable the ubuntu jobs for now

jacksavage committed:
Make the build container job an optional dep

jacksavage committed:
Add verbose flag

jacksavage committed:
Remove manual trigger from container build job

This always adds the job to the pipeline and blocks - ignoring the intent of it being an optional dependency downstream. I suppose if we want to rebuild the container, you could always go back to the last pipeline where it ran and rerun. This should be the last place that any files relevant to the container build changed.

jacksavage committed:
Use a heredoc to populate docker config

jacksavage committed:
Dump environment variables

jacksavage committed:
Disable rules for build container job for now

jacksavage committed:
Organize job into stages

This should make the pipeline summary icons in the GitLab UI more useful.

jacksavage committed:
Add epel repo and enable crb repo

For python3-mpi4py-openmpi and its deps

jacksavage committed:
Remove troubleshooting commands

jacksavage committed:
Normalize ENV declarations

jacksavage committed:
Formatting changes

jacksavage committed:
Add mpicc to path after openmpi-devel install

jacksavage committed:
Reenable ubuntu build

jacksavage committed:
Add mpicc to path if on RHEL

jacksavage committed:
Remove path mod for mpicc

jacksavage committed:
Load mpi module for RHEL

jacksavage committed:
Load mpi module instead of altering PATH

jacksavage committed:
Add boost-openmpi package to rocky linux

jacksavage committed:
Remove rocky linux jobs for now

jacksavage committed:
Bump up timeout

jacksavage committed:
Bump the test timeout again

wperkins commented on install_environment_packages.sh:
```diff
@@ -0,0 +1,42 @@
+  dnf install epel-release --assumeyes
+  crb enable
+
+  # install required packages
+  dnf install --assumeyes \
+    wget @development git python3.11 python3-pip openmpi-devel cmake pkgconf \
+    python3-mpi4py-openmpi boost-openmpi
```
You should not need to install `boost-openmpi` if you are building boost from source.  Or, you don't need to build boost from source if the system packages work.

jacksavage replied:
gotcha. I just saw the note at the end of [your comment](https://github.com/GridOPTICS/GridPACK/pull/173#issuecomment-1908532566) on the other PR. I will give that a shot. thanks.

wperkins commented on install_gridpack.sh:
```diff
@@ -1,103 +1,139 @@
+
+  # load module related to mpi4py
+  case $distribution in
+    fedora | rhel | centos | rocky)
+      echo "Loading mpi4py module"
+      module load "mpi/openmpi-$(arch)"
+      ;;
```
Just WRT the echoed message: the `module` command just puts the MPI compiler wrappers on the $PATH.  mpi4py was handled when the `python3-mpi4py-openmpi` package was installed.  

This should have been done before Boost was built too.  

Note that the `module` command only affects the current shell.   If you only want do do it once, it really needs to be done in the top-most shell (`Dockerfile`).

jacksavage replied:
I've updated the message

wperkins commented on install_gridpack_deps.sh:
```diff
@@ -1,133 +1,168 @@
+    echo "$ID"
+  )
+  case $distribution in
+  fedora | rhel | centos | rocky)
+    echo "Loading mpi4py module"
+    module load "mpi/openmpi-$(arch)"
+    ;;
```
As above, the echoed message is misleading.  `module` doesn't affect mpi4py.

jacksavage replied:
Fixed

wperkins commented on install_gridpack_deps.sh:
```diff
@@ -1,133 +1,168 @@
-  echo 'using mpi ;' >> project-config.jam
+  pushd boost || exit
 
+  # bootstrap
+  echo "Bootstrapping Boost"
+  ./bootstrap.sh \
+    --prefix=install_for_gridpack \
+    --with-libraries=mpi,serialization,random,filesystem,system
+  echo 'using mpi ;' >>project-config.jam
```
For RHEL/Rocky, the `openmpi` module needs to be loaded before this.

jacksavage replied:
Fixed

wperkins commented on install_gridpack_deps.sh:
```diff
@@ -1,133 +1,168 @@
+    --download-parmetis \
+    --download-suitesparse \
+    --download-f2cblaslapack \
+    --download-cmake \
```
Why can't PETSc use the same system CMake as GridPACK does?  Consider testing without `--download-cmake`.

jacksavage replied:
This worked without `--download-cmake`

wperkins reviewed (approved):
It looks like my comments mostly just affect the RHEL/Rocky Linux build, which can be fixed later.  I approve this PR.

jacksavage committed:
Remove cmake download flag

We're installing from source

jacksavage committed:
Remove boost-openmpi package

jacksavage committed:
Remove mpi module load from this script

jacksavage committed:
More specifically remove scripts

jacksavage committed:
Formatting

jacksavage committed:
Force remove tar file

Don't confirm ever

jacksavage committed:
Use MAKE_JOBS for global array install too

jacksavage committed:
Remove mpi module load from install_gridpack_deps

jacksavage committed:
Add script to load mpi module on RHEL

jacksavage committed:
Load the mpi module via script before bootstrapping boost

jacksavage committed:
Group scripts and dockerfile under build dir

jacksavage committed:
Move MAKE_JOBS var to global scope

jacksavage committed:
Move some variables to the devops.pnnl.gov GitLab CI/CD vars

- https://devops.pnnl.gov/groups/gridpack-code/-/settings/ci_cd
- https://devops.pnnl.gov/gridpack-code/GridPACK/-/settings/ci_cd

jacksavage committed:
Point kaniko args to new build dir

jacksavage committed:
Add open MPI vars to silence warnings

jacksavage committed:
Script renames

jacksavage committed:
Load mpi module via script before generating make files

jacksavage committed:
Update paths for kaniko args

jacksavage committed:
Test

jacksavage committed:
Rework module load

jacksavage committed:
Avoid extra chmod step

jacksavage committed:
Dot source since using sh not bash

jacksavage committed:
Put open MPI env vars in right spot to suppress warnings

jacksavage committed:
Add back script to check if container tag exists

jacksavage committed:
Add comment on bash options

jacksavage committed:
Reorder things

jacksavage committed:
Simplify arg checks

jacksavage committed:
Roll functions into lib.sh

jacksavage committed:
Check commands are installed

jacksavage committed:
Split apart check_build_container into two jobs

We don't easily have access to curl and jq in the kaniko container.

jacksavage committed:
Remove doc comments

jacksavage committed:
Use bash in dockerfile

wperkins commented on build/install_from_source_deps.sh:
```diff
@@ -0,0 +1,154 @@
+echo "Installing GridPACK dependencies"
+date
+
+install_boost "${BOOST_VERSION:?}"
+install_ga "${GA_VERSION:?}"
+install_petsc "${PETSC_VERSION:?}"
```
Reasonable defaults should be supplied here.  The script should be usable without setting a bunch of environment variables.

jacksavage replied:
I've added the defaults back inline

jacksavage committed:
Reorg scripts

jacksavage committed:
Use lib function to load mpi module and test distro

jacksavage committed:
Load mpi module before running tests

jacksavage committed:
Load mpi module from the boost script directly

Remove redundancy from dockerfile

jacksavage committed:
Add container check jobs to a stage

jacksavage committed:
Use image with bash to run container image checks

jacksavage committed:
Install jq for container image checks

jacksavage committed:
Need curl too

jacksavage committed:
Troubleshoot url issue

jacksavage committed:
Merge form data with url so theres not a false arg

jacksavage committed:
Try different header name

jacksavage committed:
Use PRIVATE_TOKEN until gitlab updated to allow JOB_TOKEN by default

jacksavage committed:
Fix newline

jacksavage committed:
Use intermediate variable between curl and jq

jacksavage committed:
Try echoing to jq

jacksavage committed:
Remove the troubleshooting from earlier...

jacksavage committed:
Use proper tag when checking registry

jacksavage committed:
Only depend on container-needs-built file in .build-container

jacksavage committed:
Try project path, not project path slug

jacksavage committed:
Inline build_container function

Missing bash causes issues sourcing lib script

jacksavage committed:
Uppercase variable

jacksavage committed:
Break script into multiple array items

jacksavage committed:
Try a different yaml string format for kaniko command

jacksavage committed:
Add a brief readme

jacksavage committed:
Add rules for when GridPACK build/test jobs should run

jacksavage committed:
Separate GP_EXT_DEPS for ref by LD_LIBRARY_PATH

jacksavage committed:
Add rocky linux jobs back

jacksavage committed:
Load mpi for GA and PETSC, fix docker RUN error handling

jacksavage committed:
Tee to build log with pipefail reporting nonzero exit from kaniko

jacksavage committed:
Provide option to run jobs from web UI

jacksavage committed:
Remove nounset for kaniko job

jacksavage committed:
Add more specific file rules for check-container-definition-changed

jacksavage committed:
Rename build to build-system, update readme contents

jacksavage committed:
Disable rockylinux jobs from automatically running for now

jacksavage committed:
Update readme

jacksavage commented:
The Rocky Linux jobs in `/.gitlab-ci.yml` have been virtually disabled using `when: manual`. This means they will only be run if someone clicks a button from the pipeline view in the GitLab UI. As the support for Rocky Linux build automation stands, an error is encountered when trying to install GridPACK to the container environment after running `/usr/bin/python3 setup.py build`. This can be seen in job [#165811](https://devops.pnnl.gov/gridpack-code/GridPACK/-/jobs/165811) on GitLab.

![image](https://github.com/GridOPTICS/GridPACK/assets/8999349/ef3b25cc-124e-4d38-a1ef-f4ef2cb71dad)

jacksavage commented:
@wperkins and @abhyshr, I think this is ready to merge and includes a decent number of changes since last review. Could you please give it another look when you have a moment?

wperkins commented:
> The Rocky Linux jobs in `/.gitlab-ci.yml` have been virtually disabled using `when: manual`. This means they will only be run if someone clicks a button from the pipeline view in the GitLab UI. As the support for Rocky Linux build automation stands, an error is encountered when trying to install GridPACK to the container environment after running `/usr/bin/python3 setup.py build`. This can be seen in job [#165811](https://devops.pnnl.gov/gridpack-code/GridPACK/-/jobs/165811) on GitLab.
> 
> [...]

This looks like the Python development package (`python3.11-devel`, maybe) was not installed.

wperkins commented on install_gridpack_deps.sh:
I'm concerned that this script no longer stands alone. Somewhere in the documentation, new users were directed to use this script to install GridPACK dependencies.  Does it work outside of the container environment?

wperkins commented on install_gridpack.sh:
Has this been tested outside of the container environment?

wperkins reviewed (commented):
I think this is fine, but, the two scripts that were deleted still need to function outside of the container environment.  These were sort of generic before. It would be good to retain that and still work on an OS other than the containers used (Debian, e.g).
